json_rpc/src/error.rs

```diff
@@ -0,0 +1,273 @@
+/// use casper_json_rpc::ErrorCodeT;
+///
+/// #[derive(Copy, Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
+/// #[repr(i64)]
+/// pub enum ErrorCode {
+///     /// The requested item was not found.
+///     NoSuchItem = -1,
+///     /// Failed to put the requested item to storage.
+///     FailedToPutItem = -2,
+/// }
+///
+/// impl From<ErrorCode> for (i64, &'static str) {
+///     fn from(error_code: ErrorCode) -> Self {
+///         match error_code {
+///             ErrorCode::NoSuchItem => (error_code as i64, "No such item"),
+///             ErrorCode::FailedToPutItem => (error_code as i64, "Failed to put item"),
+///         }
+///     }
+/// }
+///
+/// impl ErrorCodeT for ErrorCode {}
+/// ```
+pub trait ErrorCodeT:
+    Into<(i64, &'static str)> + for<'de> Deserialize<'de> + Copy + Eq + Debug
+{
+    /// Whether this type represents reserved error codes or not.
+    ///
+    /// This should normally be left with the default return value of `false`.
+    #[doc(hidden)]
+    fn is_reserved() -> bool {
+        false
+    }
+}
+
+/// The various reserved codes which can be returned in the JSON-RPC response's "error" object.
+///
+/// See [the JSON-RPC Specification](https://www.jsonrpc.org/specification#error_object) for further
+/// details.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Serialize, Deserialize, Debug)]
+#[repr(i64)]
+pub enum ReservedErrorCode {
+    /// Invalid JSON was received by the server.
+    ParseError = -32700,
+    /// The JSON sent is not a valid Request object.
+    InvalidRequest = -32600,
+    /// The method does not exist or is not available.
+    MethodNotFound = -32601,
+    /// Invalid method parameter(s).
+    InvalidParams = -32602,
+    /// Internal JSON-RPC error.
+    InternalError = -32603,
+}
+
+impl From<ReservedErrorCode> for (i64, &'static str) {
+    fn from(error_code: ReservedErrorCode) -> Self {
+        match error_code {
+            ReservedErrorCode::ParseError => (error_code as i64, "Parse error"),
+            ReservedErrorCode::InvalidRequest => (error_code as i64, "Invalid Request"),
+            ReservedErrorCode::MethodNotFound => (error_code as i64, "Method not found"),
+            ReservedErrorCode::InvalidParams => (error_code as i64, "Invalid params"),
+            ReservedErrorCode::InternalError => (error_code as i64, "Internal error"),
+        }
+    }
+}
+
+impl ErrorCodeT for ReservedErrorCode {
+    fn is_reserved() -> bool {
+        true
+    }
+}
+
+/// An object suitable to be returned in a JSON-RPC response as the "error" field.
+///
+/// See [the JSON-RPC Specification](https://www.jsonrpc.org/specification#error_object) for further
+/// details.
+#[derive(Clone, Eq, PartialEq, Serialize, Deserialize, Debug)]
+#[serde(deny_unknown_fields)]
+pub struct Error {
+    /// A number that indicates the error type that occurred.
+    code: i64,
+    /// A short description of the error.
+    message: Cow<'static, str>,
```

Comment:
Can't we use plain `&' static str` here? As far as I can see, we're always using `Cow::Borrowed`.

Reply:
I did look at that initially, but the `'static` lifetime becomes an issue when we derive `Deserialize`.  I could have worked around this by manually implementing the `Deserialize` trait, but it would have been a _lot_ of boilerplate for little to no gain I think.

Reply:
Thanks.